object_tracker/object_tracker.go

```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
 	if err := t.Reconfigure(ctx, deps, conf); err != nil {
 		return nil, err
 	}
-
 	// Default value for frequency = 10Hz
 	if t.frequency == 0 {
 		t.frequency = DefaultMaxFrequency
 	}
 
-	// Populate the first set of 2 detections to start us off
-	stream, err := t.cam.Stream(ctx, nil)
+	cancelableCtx, cancel := context.WithCancel(context.Background())
+	t.cancelFunc = cancel
+	t.cancelContext = cancelableCtx
+
+	// Do the first pass to populate the first set of 2 detections.
+	starterDets := make([][]objdet.Detection, 2)
+	stream, err := t.cam.Stream(t.cancelContext, nil)
 	if err != nil {
 		return nil, err
 	}
 	for i := 0; i < 2; i++ {
-		img, _, err := stream.Next(ctx)
+		img, _, err := stream.Next(t.cancelContext)
 		if err != nil {
 			return nil, err
 		}
-		detections, err := t.detector.Detections(ctx, img, nil)
+		detections, err := t.detector.Detections(t.cancelContext, img, nil)
 		if err != nil {
 			return nil, err
 		}
-		t.oldDetections[i] = detections
+		starterDets[i] = detections
 	}
-
-	filteredOld := FilterDetections(t.chosenLabels, t.oldDetections[0])
-	filteredNew := FilterDetections(t.chosenLabels, t.oldDetections[1])
-
-	// Rename from scratch
-	// The strategy is to rename the (n-1) detections and then match the (n) dets to those
-	// When a (n) det matches a (n-1) det, it copies its label
+	filteredOld := FilterDetections(t.chosenLabels, starterDets[0])
+	filteredNew := FilterDetections(t.chosenLabels, starterDets[1])
+	// Rename (from scratch)
 	renamedOld := make([]objdet.Detection, 0, len(filteredOld))
 	for _, det := range filteredOld {
 		newDet := t.RenameFirstTime(det)
 		renamedOld = append(renamedOld, newDet)
 	}
-
 	// Build and solve cost matrix via Munkres' method
 	matchMtx := t.BuildMatchingMatrix(renamedOld, filteredNew)
 	HA, err := hg.NewHungarianAlgorithm(matchMtx)
 	if err != nil {
 		return nil, err
 	}
 	matches := HA.Execute()
-
 	// Rename from temporal matches. New det copies old det's label
 	renamedNew := t.RenameFromMatches(matches, renamedOld, filteredNew)
-	t.oldDetections[0] = renamedOld
-	t.oldDetections[1] = renamedNew
+	t.oldDetections.Store(&[2][]objdet.Detection{renamedOld, renamedNew})
```

Comment:
You need to be careful here with shared memory -- created a `mutex` object in the struct, and every time you are about to read or write from these detections, mutexLock/Unlock around it.